nautobot/core/views/mixins.py

```diff
@@ -101,7 +101,7 @@ def get_required_permission(self):
             )
         return self.get_permissions_for_model(queryset.model, permissions)
 
-    def has_permission(self):
+    def check_permissions(self, request):
```

Comment:
Is this renamed function now being used in the same sense as DRF? Looking at the views.py you linked, it looks like the base `check_permissions` is taking action (calling `self.permission_denied()`), not returning a boolean. What's the reason for renaming this?

Reply:
There doesn't even appear to be a `has_permission` method on any DRF views. So the correct path would be merging the functionality of the prior iteration into `check_permissions`, to have it call `self.permission_denied()` vs. returning a boolean.

Reply:
Reference:

```python
    def check_permissions(self, request):
        """
        Check if the request should be permitted.
        Raises an appropriate exception if the request is not permitted.
        """
        for permission in self.get_permissions():
            if not permission.has_permission(request, self):
                self.permission_denied(
                    request,
                    message=getattr(permission, 'message', None),
                    code=getattr(permission, 'code', None)
                )
```

Reply:
Hm. `self.permission_denied()` [raises an exception](https://github.com/encode/django-rest-framework/blob/master/rest_framework/views.py#L169) in DRF, which I'm not sure we have the correct handling in place for at this time?

Reply:
It should be handled by default as it raises the `PermissionDenied` exception which results in a 403.

Reply:
like this?
```
def check_permissions(self, request):
        """
        Check whether the user has the permissions needed to perform certain actions.
        """
        user = self.request.user
        queryset = self.get_queryset()
        permission_required = self.get_required_permission()
        # Check that the user has been granted the required permission(s).
        if user.has_perms(permission_required):

            # Update the view's QuerySet to filter only the permitted objects
            for permission in permission_required:
                action = resolve_permission(permission)[1]
                queryset = queryset.restrict(user, action)
        else:
            self.permission_denied(
                request,
                message=getattr(permission_required, "message", None),
                code=getattr(permission_required, "code", None),
            )
```

Reply:
and in dispatch:
    ```
def dispatch(self, request, *args, **kwargs):
        """
        Override the default dispatch() method to check permissions first.
        Used to determine whether the user has permissions to a view and object-level permissions.
        Using AccessMixin handle_no_permission() to deal with Object-Level permissions and API-Level permissions in one pass.
        """
        # initialize_request also instantiates self.action which is needed for permission checks.
        self.initialize_request(request, *args, **kwargs)
        try:
            self.check_permissions(request)
        except:
            return self.handle_no_permission()

        return super().dispatch(request, *args, **kwargs)

```

Reply:
In the former, the `has_perms` stuff looks accurate. In the `dispatch` method, the only thing I don't like ks the naked `except`.

I am realizing we have a little confusion here, which should be clarified and documented for posterity.

1. My comment about `self.permission_denied()` raising `PermissionDenied` exception is the **DRF version of this exception** which when raised will return a JSON 403 DRF `Response` object. See: https://www.django-rest-framework.org/api-guide/exceptions/#permissiondenied

2. The method `self.handle_no_permission()` *optionally* raises Django's version of `PermissionDenied` if `raise_exception = True` is set on the view class (if it inherits from `AccessMixin`), otherwise it will d a login redirect. See: https://docs.djangoproject.com/en/4.1/ref/exceptions/#permissiondenied

Reply:
I tried to catch the DRF version of 403 `PermissionDenied` exception by `self.permission_denied()`. However, before that exception is even raised I ran into an `AttributeError `which is similar to what is explained [here](https://github.com/encode/django-rest-framework/issues/7210). The request.authenticator is a part of `self.permission_denied()` and is only accessible in API requests apparently. 
I think we can either stick with my current approach just to catch all exceptions and return `self.handle_no_permission()` (which I dont like it either) or we can revert to my previous approach: instead of raising an DRF exceptions we can return True or False from `self.check_permissions()`. What do you think? 🤔

Reply:
Ahh, thanks for that context. Honestly since it seems like we'll be expecting this exception every time, we could just catch both `AttributeError` and `PermissionDenied` to split the difference and I'd be happy. Can you also please provide a comment linking to https://github.com/encode/django-rest-framework/issues/7210?

Reply:
Okay so apparently I can convert a WSGI request to an API request by doing `api_request = self.initialize_request(request, *args, **kwargs)` and pass `api_request` to self.check_permissions() instead. Now DRF's `exceptions.NotAuthenticated, exceptions.PermissionDenied` can be raised properly and handled by `self.handle_no_permission` in the except block. Check out this [commit](https://github.com/nautobot/nautobot/pull/2598/commits/4f98e27b0436271a817a5257958a5500d041e27b) to see the change I made. I also added some comments.